File: terminal/src/com/jediterm/terminal/emulator/SystemCommandSequence.java
package com.jediterm.terminal.emulator;

import com.jediterm.terminal.TerminalDataStream;
import com.jediterm.terminal.util.CharUtils;
import com.jediterm.typeahead.Ascii;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @author traff
 */
final class SystemCommandSequence {

  private static final char ST = 0x9c;

  private final List<Object> myArgs = new ArrayList<>();
  private final StringBuilder mySequence = new StringBuilder();

  public SystemCommandSequence(@NotNull TerminalDataStream stream) throws IOException {
    StringBuilder argBuilder = new StringBuilder();
    boolean end = false;
    while (!end) {
      char ch = stream.getChar();
      mySequence.append(ch);
      end = isEnd();
      if (ch == ';' || end) {
        if (end && isTwoBytesEnd()) {
          argBuilder.deleteCharAt(argBuilder.length() - 1);
        }
        myArgs.add(parseArg(argBuilder.toString()));
        argBuilder.setLength(0);
      }
      else {
        argBuilder.append(ch);
      }
    }
  }

  private @NotNull Object parseArg(@NotNull String arg) {
    if (arg.length() > 0 && Character.isDigit(arg.charAt(arg.length() - 1))) {
      // check isDigit to reduce amount of expensive NumberFormatException
      try {
        return Integer.parseInt(arg);
      }
      catch (NumberFormatException ignored) {
      }
    }
    return arg;
  }

  private boolean isEnd() {
    int len = mySequence.length();
    if (len > 0) {
      char ch = mySequence.charAt(len - 1);
      return ch == Ascii.BEL || ch == ST || isTwoBytesEnd();
    }
    return false;
  }

  private boolean isTwoBytesEnd() {
    int len = mySequence.length();
    return len > 1 && mySequence.charAt(len - 2) == Ascii.ESC && mySequence.charAt(len - 1) == '\\';
  }

  public @Nullable String getStringAt(int i) {
    if (i>=myArgs.size()) {
      return null;
    }
    Object val = myArgs.get(i);
    return val instanceof String ? (String)val : null;
  }

  public int getIntAt(int position, int defaultValue) {
    if (position < myArgs.size()) {
      Object val = myArgs.get(position);
      if (val instanceof Integer) {
        return (Integer) val;
      }
    }
    return defaultValue;
  }

  public @NotNull String format(@NotNull String body) {
    return (char)Ascii.ESC + "]" + body + getTerminator();
  }

  @Override
  public String toString() {
    return CharUtils.toHumanReadableText(mySequence.toString());
  }

  /**
   * <a href="https://invisible-island.net/xterm/ctlseqs/ctlseqs.html">
   * XTerm accepts either BEL or ST for terminating OSC
   * sequences, and when returning information, uses the same
   * terminator used in a query. </a>
   */
  private @NotNull String getTerminator() {
    int len = mySequence.length();
    if (isTwoBytesEnd()) {
      return mySequence.substring(len - 2);
    }
    return mySequence.substring(len - 1);
  }
}

File: bossterm-core-mpp/src/jvmMain/kotlin/ai/rever/bossterm/terminal/emulator/SystemCommandSequence.kt
package ai.rever.bossterm.terminal.emulator

import ai.rever.bossterm.core.util.Ascii
import ai.rever.bossterm.terminal.TerminalDataStream
import ai.rever.bossterm.terminal.util.CharUtils
import java.io.IOException
import java.lang.StringBuilder

internal class SystemCommandSequence
@Throws(IOException::class) constructor(stream: TerminalDataStream) {

  private val text: String
  val args: List<String>

  init {
    val textBuf = StringBuilder()
    // Large limit to support inline images (OSC 1337;File)
    // A 10MB image = ~13MB base64 = 13 million characters
    val maxLength = 50 * 1024 * 1024  // 50MB max for inline images

    do {
      if (textBuf.length >= maxLength) {
        throw IOException("OSC sequence exceeds maximum length of $maxLength characters")
      }

      val ch = stream.char
      if (ch == Ascii.NUL) {
        throw IOException("Unexpected end of stream in OSC sequence")
      }

      textBuf.append(ch)
    } while (!isTerminated(textBuf))

    text = textBuf.toString()
    val body = text.dropLast(terminatorLength(text))
    args = java.util.List.copyOf(body.split(ARG_SEPARATOR))
  }

  fun getStringAt(index: Int): String? = args.getOrNull(index)

  fun getIntAt(index: Int, defaultValue: Int): Int = parseInt(args.getOrNull(index), defaultValue)

  private fun parseInt(value: String?, defaultValue: Int): Int {
    return value?.toIntOrNull() ?: defaultValue
  }

  fun format(args: List<String>): String {
    // https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h3-Operating-System-Commands
    // XTerm accepts either BEL  or ST  for terminating OSC
    // sequences, and when returning information, uses the same
    // terminator used in a query.
    return Ascii.ESC + "]" + args.joinToString(ARG_SEPARATOR.toString()) + text.takeLast(terminatorLength(text))
  }

  override fun toString(): String = CharUtils.toHumanReadableText(Ascii.ESC.toString() + "]" + text)

  internal companion object {
    private const val ST: Char = 0x9c.toChar()
    private const val ARG_SEPARATOR: Char = ';'
    internal const val OSC: Char = 0x9d.toChar() // C1 control code

    private fun isTerminated(text: CharSequence): Boolean {
      val len = text.length
      if (len > 0) {
        val ch = text[len - 1]
        return ch == Ascii.BEL || ch == ST || isTwoBytesTerminator(text, len, ch)
      }
      return false
    }

    private fun isTwoBytesTerminator(text: CharSequence, length: Int, lastChar: Char): Boolean {
      return lastChar == '\\' && length >= 2 && text[length - 2] == Ascii.ESC
    }

    private fun terminatorLength(text: CharSequence): Int {
      val length = text.length
      return if (isTwoBytesTerminator(text, length, text[length - 1])) 2 else 1
    }
  }
}